src/coreclr/src/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -507,7 +572,7 @@ private void MergeWithGlobalList(T[] list)
                             }
 
                             Debug.Assert(cachedMembers![freeSlotIndex] == null);
-                            cachedMembers[freeSlotIndex] = newMemberInfo;
+                            Volatile.Write(ref cachedMembers[freeSlotIndex], newMemberInfo); // value may be read outside of lock
```

Comment:
I always find it hard to reason about volatile. For writes "If a read or write appears before this method in the code, the processor cannot move it after this method."

But all of the relevant code that appears before this method in the code is within a lock. Could you help me understand what bad scenario is possible without the memory barrier here?

Reply:
I don't think anything bad can happen in practice due to some JIT implementation details. But _in theory_ without this volatile write one thread could publish a `MemberInfo` instance whose construction is not yet seen as complete. Normally this would be guarded by the lock, which provides a full memory barrier. But since we've introduced a code path which traverses this array outside the lock, we need the volatile write here to ensure the reader never sees a partially published instance.

Again, I don't think it's absolutely necessary since I'm pretty sure as an implementation detail the JIT always makes all reference-type writes volatile. But I don't know if we actually say that as an ironclad guarantee anywhere, so seems better to be safe here.

Reply:
I see.